agent/ecscni/plugin.go

```diff
@@ -110,92 +109,115 @@ func (client *cniClient) SetupNS(cfg *Config) (*current.Result, error) {
-	return client.libcni.DelNetworkList(networkConfigList, cns)
+	err = client.eniDel(runtimeConfig, cfg)
+	if err != nil {
+		return errors.Wrap(err, "cni cleanup: invoke eni plugin failed")
+	}
+	seelog.Debugf("ecscni: container namespace cleanup done: %s", cfg.ContainerID)
+	return nil
 }
 
 // ReleaseIPResource marks the ip available in the ipam db
 func (client *cniClient) ReleaseIPResource(cfg *Config) error {
-	cns := &libcni.RuntimeConf{
+	runtimeConfig := &libcni.RuntimeConf{
 		ContainerID: cfg.ContainerID,
 		NetNS:       fmt.Sprintf(netnsFormat, cfg.ContainerPID),
-		IfName:      defaultEthName,
+		IfName:      defaultVethName,
 	}
 
 	ipamConfig, err := client.createIPAMNetworkConfig(cfg)
 	if err != nil {
 		return err
 	}
 
-	networkConfigList := &libcni.NetworkConfigList{
-		CNIVersion: client.cniVersion,
-		Plugins:    []*libcni.NetworkConfig{ipamConfig},
-	}
-
 	seelog.Debugf("Releasing the ip resource from ipam db, id: [%s], ip: [%v]", cfg.ID, cfg.IPAMV4Address)
 	os.Setenv("ECS_CNI_LOGLEVEL", logger.GetLevel())
 	defer os.Unsetenv("ECS_CNI_LOGLEVEL")
-	return client.libcni.DelNetworkList(networkConfigList, cns)
+	return client.libcni.DelNetwork(ipamConfig, runtimeConfig)
 }
 
-// createNetworkConfig creates the configuration for invoking plugins
-func (client *cniClient) createNetworkConfig(cfg *Config, bridgeConfigFunc func(*Config) (*libcni.NetworkConfig, error)) (*libcni.NetworkConfigList, error) {
-	bridgeConfig, err := bridgeConfigFunc(cfg)
+// bridgeAdd invokes the bridge and ipam plugin to set up the bridge,
+// also it sets the veth pair and route in container namespace
+func (client *cniClient) bridgeAdd(runtimeConfig libcni.RuntimeConf, cfg *Config) (cnitypes.Result, error) {
+	deviceName, networkConfig, err := client.createBridgeNetworkConfigWithIPAM(cfg)
 	if err != nil {
 		return nil, err
 	}
+	runtimeConfig.IfName = deviceName
 
-	eniConfig, err := client.createENINetworkConfig(cfg)
+	return client.libcni.AddNetwork(networkConfig, &runtimeConfig)
+}
+
+// eniAdd invokes the eni plugin to setup the eni in container namespace
+func (client *cniClient) eniAdd(runtimeConfig libcni.RuntimeConf, cfg *Config) (cnitypes.Result, error) {
+	deviceName, networkConfig, err := client.createENINetworkConfig(cfg)
 	if err != nil {
 		return nil, err
 	}
 
-	// We order bridgeConfig after eniConfig so that the result of the bridge
-	// plugin (which contains results of the eni plugin) can be parsed to
-	// get the IP address of the veth pair created
-	pluginConfigs := []*libcni.NetworkConfig{eniConfig, bridgeConfig}
-	networkConfigList := &libcni.NetworkConfigList{
-		CNIVersion: client.cniVersion,
-		Plugins:    pluginConfigs,
+	runtimeConfig.IfName = deviceName
+	return client.libcni.AddNetwork(networkConfig, &runtimeConfig)
+}
+
+// bridgeDel invokes the bridge plugin to clean up the container namespace
+func (client *cniClient) bridgeDel(runtimeConfig libcni.RuntimeConf, cfg *Config) error {
```

Comment:
Can you combine `bridgeDel` and `eniDel` into a single method similar to my other comment?